kingpin/actors/aws/ecs.py

```diff
@@ -406,7 +420,369 @@ def _execute(self):
+                                       client_token)
+        else:
+            errors = _get_immutable_field_errors(
+                old_params=existing_service,
+                new_params=self.service_definition,
+                immutable_fields=['loadBalancers', 'role'])
+
+            if errors:
+                for error in errors:
+                    self.log.error(error)
+                raise exceptions.UnrecoverableActorFailure(
+                    'Immutable fields cannot be updated. '
+                    'A new service must be created.')
+
+            yield self._update_service(service_name,
+                                       family, revision)
+        raise gen.Return()
+
+    @gen.coroutine
+    @dry('Would wait for service to deploy')
+    def _wait_for_deployment(self, service_name):
+        """Wait for service to deploy on ECS.
+
+        If the service fails to come up,
+        this will raise an exception.
+
+        Args:
+            service_name: name of the service to wait for.
+        """
+        # Used to ensure event logs are only printed once.
+        seen_events = set()
+        while True:
+            done = yield self._service_deployed(
+                service_name,
+                seen_events)
+            if done:
+                break
+            yield gen.sleep(10)
+
+    @gen.coroutine
+    def _service_deployed(self, service_name, seen_events):
+        """Checks if service is finished deploying.
+
+        Args:
+            service_name: name of the service to wait for.
+            seen_events: set of seen event ids.
+
+        Returns:
+            A boolean indicating whether the service is completely deployed.
+        """
+        service = yield self._describe_service(service_name)
+
+        deployments = service['deployments']
+        primary_deployment = None
+        for deployment in deployments:
+            if deployment['status'] == 'PRIMARY':
+                primary_deployment = deployment
+                break
+        if not primary_deployment:
+            # There should always be one 'PRIMARY' deployment returned.
+            raise exceptions.UnrecoverableActorFailure(
+                'No primary deployment.')
+
+        service_timestamp = primary_deployment['createdAt']
+
+        sorted_new_events = _get_sorted_new_log_events(
+            events=service['events'],
+            start_timestamp=service_timestamp,
+            seen_events=seen_events)
+
+        for event in sorted_new_events:
+            event_timestamp, event_message = event
+            self.log.info('Event: {message}'.format(
+                message=event_message))
+
+        running_count = primary_deployment['runningCount']
+        desired_count = primary_deployment['desiredCount']
+        missing_count = desired_count - running_count
+
+        extra_deployment_count = len(deployments) - 1
+
+        if missing_count == 0 and extra_deployment_count == 0:
```

Comment:
So I'm not sure if this is a realistic problem -- but by only checking that the number of deployments is correct, I think that we could see a race condition where the Amazon API returns stale data. I could see a situation where we call this method _right_ after we call the `_update_service()` method, and they return only the "primary" deployment and not the "new" deployment thats being set up.

Can we add some more checking here to make sure that the "primary deployment" task ARN matches what we expect?